pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/InstancePartitions.java

```diff
@@ -0,0 +1,95 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.controller.helix.core.assignment;
+
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import org.apache.helix.ZNRecord;
+
+
+/**
+ * Instance (server) partitions for the table.
+ *
+ * <p>The instance partitions are stored as a map from partition of the format: {@code <partitionId>_<replicaId>} to
+ * list of server instances.
+ */
+public class InstancePartitions {
+  private static final char SPLITTER = '_';
+
+  private final String _tableNameWithType;
+  private final Map<String, List<String>> _partitionToInstancesMap;
+  private int _numPartitions;
+  private int _numReplicas;
```

Comment:
I guess that `numReplicas` for `InstancePartitions` is equivalent to `the number of groups in instance partition`. In that case, we may need to consider changing variable name for`numReplicas` to `numReplicaGroups`?  

In most cases, `numReplica` here is equal to `replication` factor; however, we use `numReplica=1` for supporting backward compatibility. In this case, `numReplica` and `replication` will mean different numbers. (`numReplica=1` while `replication can be 3`)

Reply:
This `numReplicas` is for instance partitions (when we index it with partitionId and replicaId, replicaId must be smaller than this numReplicas).
You are correct that the instance partitions **numReplicas** might not match the table "replication", but it only represent the number of replicas inside instance partitions. Inside instance partitions, there is no concept of replica group, so I would prefer keep the variable name `numReplicas`

Reply:
I agree with snlee. It is confusing to call this numReplicas. numReplicas already has a well known implied meaning, and this `numInstanceGroups`, or `numInstancePartitions` shouldn't be called numReplicas

Reply:
@npawar this `numReplicas` has the same meaning as the implied meaning. The only difference is that we use `numReplicas` **1** for backward compatibility.